x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/SchemaUtil.java

```diff
@@ -23,30 +23,59 @@
 import org.elasticsearch.xpack.core.ClientHelper;
 import org.elasticsearch.xpack.core.transform.transforms.pivot.PivotConfig;
 
+import java.math.BigDecimal;
 import java.util.HashMap;
 import java.util.Map;
 import java.util.Objects;
-import java.util.Set;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 public final class SchemaUtil {
     private static final Logger logger = LogManager.getLogger(SchemaUtil.class);
 
-    // Full collection of numeric field type strings
-    private static final Set<String> NUMERIC_FIELD_MAPPER_TYPES;
+    // Full collection of numeric field type strings and whether they are floating point or not
+    private static final Map<String, Boolean> NUMERIC_FIELD_MAPPER_TYPES;
     static {
-        Set<String> types = Stream.of(NumberFieldMapper.NumberType.values())
-            .map(NumberFieldMapper.NumberType::typeName)
-            .collect(Collectors.toSet());
-        types.add("scaled_float"); // have to add manually since scaled_float is in a module
+        Map<String, Boolean> types = Stream.of(NumberFieldMapper.NumberType.values())
+            .collect(Collectors.toMap(t -> t.typeName(), t -> t.numericType().isFloatingPoint()));
+
+        // have to add manually since they are in a module
+        types.put("scaled_float", true);
+        types.put("unsigned_long", false);
         NUMERIC_FIELD_MAPPER_TYPES = types;
     }
 
     private SchemaUtil() {}
 
     public static boolean isNumericType(String type) {
-        return type != null && NUMERIC_FIELD_MAPPER_TYPES.contains(type);
+        return type != null && NUMERIC_FIELD_MAPPER_TYPES.containsKey(type);
+    }
+
+    /**
+     * Convert a numeric value to an integer if it's not a floating point number.
+     *
+     * Implementation decision: We do not care about the concrete type, but only if its floating point or not.
+     * Further checks (e.g. range) are done at indexing.
+     *
+     * If type is floating point and value could be an integer (ends with `.0`), we still preserve `.0` in case
+     * the destination index uses dynamic mappings as well as being json friendly.
+     *
+     * @param type the type of the value according to the schema we know
+     * @param value the value as double (aggs return double for everything)
+     * @return value if its floating point, a integer
+     */
+    public static Object convertToIntegerTypeIfNeeded(String type, double value) {
```

Comment:
```suggestion
    public static Object convertToDiscreteNumeralTypeIfNeeded(String type, double value) {
```
Or some better name as this is is not an `integer` but instead some possibly large discrete numeral (big int or a long).

Reply:
its referring to the mathematical definition of integer, that's why integer _type_, not integer. I can try to find a better name, maybe `dropFloatingPointComponentIfNeeded`, but lets not over-complicate it, the doc string defines what it does.